pkg/log/logshipper/logshipper.go

```diff
@@ -192,42 +182,136 @@ func filterResults(keyvals ...interface{}) {
 	}
 }
 
-// addDeviceIdentifyingAttributesToLogger gets device identifiers from the server-provided
-// data and adds them as attributes on the logger.
-func (ls *LogShipper) addDeviceIdentifyingAttributesToLogger() {
-	additionalSlogAttrs := make([]slog.Attr, 0)
+// updateDevideIdentifyingAttributes gets device identifiers from the server-provided
+// data and adds them as attributes on the logger. If shipping has not yet started and
+// there are logs in the buffer, the device identifiers are added to the logs in the
+// buffer.
+func (ls *LogShipper) updateDevideIdentifyingAttributes() error {
+	deviceInfo := make(map[string]string)
 
 	versionInfo := version.Version()
+	deviceInfo["version"] = versionInfo.Version
 	ls.shippingLogger = log.With(ls.shippingLogger, "version", versionInfo.Version)
-	additionalSlogAttrs = append(additionalSlogAttrs, slog.String("version", versionInfo.Version))
 
-	if deviceId, err := ls.knapsack.ServerProvidedDataStore().Get([]byte("device_id")); err != nil {
-		level.Debug(ls.baseLogger).Log("msg", "could not get device id", "err", err)
-	} else {
-		ls.shippingLogger = log.With(ls.shippingLogger, "k2_device_id", string(deviceId))
-		additionalSlogAttrs = append(additionalSlogAttrs, slog.String("k2_device_id", string(deviceId)))
+	for _, key := range []string{"device_id", "munemo", "organization_id", "serial_number"} {
+		v, err := ls.knapsack.ServerProvidedDataStore().Get([]byte(key))
+		if err != nil {
+			return fmt.Errorf("could not get %s from server provided data: %w", key, err)
+		}
+
+		if len(v) == 0 {
+			return fmt.Errorf("no value for %s in server provided data", key)
+		}
+
+		deviceInfo[key] = string(v)
+	}
+
+	ls.shippingLogger = log.With(ls.shippingLogger, deviceInfo)
```

Comment:
Okay, I'm somewhat wrong -- you have to splat it out. Calling like this has the key set to `deviceInfo` and the value nil. Take a look at https://go.dev/play/p/bJMUWw28Ekp (Or building the array as you go)

Reply:
ahh, good catch, updated